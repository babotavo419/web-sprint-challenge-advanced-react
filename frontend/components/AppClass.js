import React from 'react'

const initialMessage = ''
const initialEmail = ''
const initialSteps = 0
const initialIndex = 4 

const initialState = {
  message: initialMessage,
  email: initialEmail,
  index: initialIndex,
  steps: initialSteps,
}

export default class AppClass extends React.Component {

getXY = () => {
  const x = this.state.index % 3 + 1; 
  const y = Math.floor(this.state.index / 3) + 1;
}

getXYMessage = () => {
  const [x, y] = this.getXY();
  return `Coordinates (${x}, ${y})`;
}

reset = () => {
  this.setState(initialState);
}

getNextIndex = (direction) => {
  let newIndex = this.state.index;
  switch (direction) {
    case 'left':
      if (newIndex % 3 > 0) newIndex -= 1;
      break;
    case 'up':
      if (newIndex - 3 >= 0) newIndex -= 3;
      break;
    case 'right':
      if (newIndex % 3 < 2) newIndex += 1;
      break;
    case 'down':
      if (newIndex + 3 < 9) newIndex += 3;
      break;
  }
  return newIndex;
}

move = (evt) => {
  const direction = evt.target.id;
  const newIndex = this.getNextIndex(direction);
  this.setState({ index: newIndex, steps: this.state.steps + 1 });
}

onChange = (evt) => {
  this.setState({ email: evt.target.value });
}

onSubmit = async (evt) => {
  evt.preventDefault();
  const [x, y] = this.getXY();
  try {
    const response = await axios.post('http://localhost:9000/api/result', {
      x,
      y,
      steps: this.state.steps,
      email: this.state.email,
    });
    this.setState({ message: response.data.message });
  } catch (error) {
    this.setState({ message: error.response.data.message });
  }
}

  render() {
    const { className } = this.props
    return (
      <div id="wrapper" className={className}>
        <div className="info">
          <h3 id="coordinates">Coordinates (2, 2)</h3>
          <h3 id="steps">You moved 0 times</h3>
        </div>
        <div id="grid">
          {
            [0, 1, 2, 3, 4, 5, 6, 7, 8].map(idx => (
              <div key={idx} className={`square${idx === 4 ? ' active' : ''}`}>
                {idx === 4 ? 'B' : null}
              </div>
            ))
          }
        </div>
        <div className="info">
          <h3 id="message"></h3>
        </div>
        <div id="keypad">
          <button id="left">LEFT</button>
          <button id="up">UP</button>
          <button id="right">RIGHT</button>
          <button id="down">DOWN</button>
          <button id="reset">reset</button>
        </div>
        <form>
          <input id="email" type="email" placeholder="type email"></input>
          <input id="submit" type="submit"></input>
        </form>
      </div>
    )
  }
}
