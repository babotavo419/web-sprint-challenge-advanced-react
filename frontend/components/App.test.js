import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import AppClass, { initialState } from './AppClass';

describe('<AppClass />', () => {
  let component;

  beforeEach(() => {
    component = render(<AppClass />);
  });

  it('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('shows the correct initial coordinates and steps', () => {
    expect(component.getByTestId('coordinates')).toHaveTextContent('Coordinates (2, 2)');
    expect(component.getByTestId('steps')).toHaveTextContent('You moved 0 times');
  });

  it('moves correctly to the right and updates steps and coordinates', () => {
    fireEvent.click(component.getByText('RIGHT'));
    expect(component.getByTestId('coordinates')).toHaveTextContent('Coordinates (3, 2)');
    expect(component.getByTestId('steps')).toHaveTextContent('You moved 1 time');
  });

  it('does not move when hitting the boundary', () => {
    fireEvent.click(component.getByText('UP'));
    fireEvent.click(component.getByText('UP'));
    expect(component.getByTestId('coordinates')).toHaveTextContent('Coordinates (2, 1)');
    expect(component.getByTestId('steps')).toHaveTextContent('You moved 1 time');
  });

  it('resets the game correctly', () => {
    fireEvent.click(component.getByText('RIGHT'));
    fireEvent.click(component.getByText('reset'));
    expect(component.getByTestId('coordinates')).toHaveTextContent('Coordinates (2, 2)');
    expect(component.getByTestId('steps')).toHaveTextContent('You moved 0 times');
  });
});

