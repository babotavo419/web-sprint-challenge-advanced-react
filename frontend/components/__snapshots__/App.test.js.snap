// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppClass /> renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        id="wrapper"
      >
        <div
          class="info"
        >
          <h3
            id="coordinates"
          >
            Coordinates (2, 2)
          </h3>
          <h3
            id="steps"
          >
            You moved 
            0
             time
            s
          </h3>
        </div>
        <div
          id="grid"
        >
          <div
            class="square"
          />
          <div
            class="square"
          />
          <div
            class="square"
          />
          <div
            class="square"
          />
          <div
            class="square active"
          >
            B
          </div>
          <div
            class="square"
          />
          <div
            class="square"
          />
          <div
            class="square"
          />
          <div
            class="square"
          />
        </div>
        <div
          class="info"
        >
          <h3
            id="message"
          />
        </div>
        <div
          id="keypad"
        >
          <button
            id="left"
          >
            LEFT
          </button>
          <button
            id="up"
          >
            UP
          </button>
          <button
            id="right"
          >
            RIGHT
          </button>
          <button
            id="down"
          >
            DOWN
          </button>
          <button
            id="reset"
          >
            reset
          </button>
        </div>
        <form>
          <input
            id="email"
            placeholder="type email"
            type="email"
            value=""
          />
          <input
            id="submit"
            type="submit"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      id="wrapper"
    >
      <div
        class="info"
      >
        <h3
          id="coordinates"
        >
          Coordinates (2, 2)
        </h3>
        <h3
          id="steps"
        >
          You moved 
          0
           time
          s
        </h3>
      </div>
      <div
        id="grid"
      >
        <div
          class="square"
        />
        <div
          class="square"
        />
        <div
          class="square"
        />
        <div
          class="square"
        />
        <div
          class="square active"
        >
          B
        </div>
        <div
          class="square"
        />
        <div
          class="square"
        />
        <div
          class="square"
        />
        <div
          class="square"
        />
      </div>
      <div
        class="info"
      >
        <h3
          id="message"
        />
      </div>
      <div
        id="keypad"
      >
        <button
          id="left"
        >
          LEFT
        </button>
        <button
          id="up"
        >
          UP
        </button>
        <button
          id="right"
        >
          RIGHT
        </button>
        <button
          id="down"
        >
          DOWN
        </button>
        <button
          id="reset"
        >
          reset
        </button>
      </div>
      <form>
        <input
          id="email"
          placeholder="type email"
          type="email"
          value=""
        />
        <input
          id="submit"
          type="submit"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
